### PR TITLE
fix: we should still return tags pointing to deleted branches

### PR DESCRIPTION
**Pull request by bcoe**
we should still find semver git tags that point to branches (addresses https://github.com/conventional-changelog/standard-version/issues/123, https://github.com/conventional-changelog/standard-version/issues/126)

We should also make sure that we always sort chronologically, addressed by `--date-order`.

I've tested this functionality thoroughly [here](https://github.com/CHANGELOG-bot/debug-standard-version) and have been unable to recreate the issues that I was running into.

CC: @Tapppi, @stevemao

### CODE REVIEW

**Comment by AntJanus** [comment]
Okay, so this is really confusing to me. And I wonder if you can clarify things for me. I've been using git-flow-like workflow in git, meaning that I have a `development` branch and I always create a `release` branch. A month or so ago, I started doing manual semver. The release branch was always merged in via Github using the standard merge method.

A week ago, I started using `standard-version` which recognized my previous semver tags; however, this time, I merged the branch in via squash (or possibly rebase) using Github. When I run `git-semver-tags`, only my manually-made tags show up, not the tag made by `standard-version`.

What's the difference between squash/rebase/merge such that one of the tags becomes invisible? And more importantly, until this fix is merged in, what can I do to remedy the situation manually?

**Comment by bcoe** [comment]
@AntJanus in my case, I was occasionally releasing a candidate release from a branch which I would then squash. So the tag was pointing at a git-sha that did exist on the history reported by: `git log`. By running `git log --date-order --all`, a history is returned with both the squash commit and the commits that were squashed.

Can you confirm that this patch also solves your issue?

**Review by stevemao — APPROVED** [review]
Wow, I never thought it could be solved by this :) thanks @bcoe !

**Review by stevemao — CHANGES_REQUESTED** [review]
Can we add this to readme?

**Comment by stevemao** [comment]
Just a thought. Do you think we should include this as an option?

**Comment by bcoe** [comment]
The thought had crossed my mind but, unless I'm missing something, I think
this is always probably the best behavior? Can't think of a scenario where
I'd want to ignore tags.

**Review by stevemao — APPROVED** [review]

**Comment by AntJanus** [comment]
Looks like I had some other issues as well that caused repeating entries (always squash branches into master, everyone!) but it worked! Thanks! :)

**Comment by AntJanus** [comment]
I should clarify, I went into the node_modules and adjusted `git-semver-tags` and ran `standard-version` to see if it would generate the changelog correctly. There were some issues but unrelated to this problem. I ended up adjusting the globally installed package and running `git-semver-tags` and got the right tag listing.

**Comment by bcoe** [comment]
@AntJanus awesome, for future reference, this is a lovely tool for this sort of task:

https://docs.npmjs.com/cli/link

**Comment by AntJanus** [comment]
@bcoe I knew there was a better way but honestly, it was one line and I wanted to test it out in my specific repo as quickly as possible.

:) Thanks for the tip!

**Comment by AntJanus** [comment]
Is there a timeline for releases btw? I just want to know if I should just apply the manual patch for now and use that or wait

**Comment by Tapppi** [comment]
So this isn't as simple as one might think. I thought up some side effects with test/pre-releases on other branches. Say you do the following:
- A stable release on `master`
- Merge/do changes on `master`
- Branch and do changes on new feature branch `awsum`
- Release a pre-release on `awsum`
- Never merge `awsum` because it was actually not very awsome, and continue work on `master`
- Do a release on `master` after some more changes

Unless you delete your `awsum` pre-release **tag** (deleting branch is not enough) manually, changes on `master` that happened after your last release, but before the initial branch-off of `awsum` will never get added to your changelog on master. But if you DO delete the tag on branch `awsum`, your next bump will potentially be incorrect, because there is a semver tag missing! This is because both the _bump_  and the _changelog_  portion call `git-semver-tags` with default options. I tested this by doing changes on your nice test project @bcoe, thanks for that!

This is a very complicated workflow (I personally don't allow releases off any other than the main release branch, i.e. `master` because of potential problems such as these). One option I see is to make this an option and have `git-latest-semver-tags` (used by `conventional-recommended-bump`) use that option while `conventional-changelog-core` would not. That might bring up new edge cases though..

Also btw, makes no sense that this package is under `conventional-changelog` org but https://github.com/stevemao/git-latest-semver-tag is not, could you transfer the project @stevemao? 😄

**Comment by bcoe** [comment]
@Tapppi I think this approach gets us 95% of the way there; I definitely have the tendency to sometimes tag a long-lived feature branch, if I know it's about to land and I want to publish a package for the community to test -- I find it's much easier to tell people to run:

`npm cache clear; npm i my-module@next` 

than to have them clone a git repository and link it. Seems like mainly you'd like to see this become a configuration option, that's disabled by default -- I'm fine with this, but think we should enable it for `standard-version`.

**Comment by bcoe** [comment]
@Tapppi noting that I've bit myself three or four times with this issue at this point, and I think in each case I would have been saved by this setting.

**Comment by Tapppi** [comment]
@bcoe Sure, I think this is the right step forward, but maybe make it an option here (enabled in `standard-version`, might have to propagate the option somewhere). I'd prefer this be documented in `standard-version` as well, but maybe I can pitch that in. People might have no idea how easily something can go awry if you don't merge.

The issues I outlined above arise as soon as you tag another branch, but then don't merge it, so I think it's a valid concern.

**Comment by gkatsev** [comment]
I'm running into this issue myself with videojs. To support bower and to keep our git history clean, we version and tag separately. We version in master but then we create a temporary branch where we build our distfiles, force-commit them into git, tag there, and then delete the branch.

I have a question, though. Why not use `git tag` to get a list of all tags available and then filter out the ones that aren't semver? Also, should somewhere down-the-line, perhaps conventional-changelog, take into account the version in package.json?

I found a potential issue as well. Say, you have a 1.0.0 release and a 1.1.0 release but then you find a bug in 1.0.0 that you want to address, so, you fix it in 1.0.1. Now because the logs are date-ordered, we end up thinking that 1.0.1 is the latest release instead of 1.1.0.

---

Also, I just wanted to thank y'all for the great software!

**Comment by stevemao** [comment]
> Why not use git tag to get a list of all tags available and then filter out the ones that aren't semver

[It does](https://github.com/conventional-changelog/git-semver-tags/blob/master/index.js#L22-L24)

> Say, you have a 1.0.0 release and a 1.1.0 release but then you find a bug in 1.0.0 that you want to address, so, you fix it in 1.0.1. Now because the logs are date-ordered, we end up thinking that 1.0.1 is the latest release instead of 1.1.0.

Isn't this expected? See npm's changelog. they release 2 and 3 in parallel.

**Comment by gkatsev** [comment]
Let me clarify, I mean [this](https://github.com/conventional-changelog/git-semver-tags/blob/master/index.js#L5) command. So, instead, run `git tag` which will return a list of the tags available in the current repo.

But my issue is that the list of tags comes out ordered like so:

```
1.0.1
1.1.0
1.0.0
```

But I'm in the process of releasing 1.2.0 or 1.1.1 and so, it thinks that my latest version is 1.0.1, so, the changelog includes all the changes that have been included after 1.0.1, including changes in 1.1.0 but I only want changes in 1.2.0.

I can elaborate more if it's still not clear. Also, let me know if you want me to move this into a separate issue.

**Comment by gkatsev** [comment]
The main issue with `git tag` is that the list it returns is in alphabetical ordering, so, `1.12` would come before `1.2` but that can easily re-sorted.

**Comment by stevemao** [comment]
@gkatsev Maybe you need to do something on your branches so that the tool can pick the right commits for those releases, and do the right releases on those branches. You'd need to carefully merge them.
Because if you didn't have a changelog, by just using `git log` I'd be confused too.
Unless you have a better idea?

**Comment by stevemao** [comment]
> so, 1.12 would come before 1.2 but that can easily re-sorted.

this doesn't work if you release v2 and v3 in parallel.

**Comment by gkatsev** [comment]
The git log does have a commit indicating a new version, so, if you're just reading the git history it looks good. The reason we commit on a branch that isn't pushed up is because we want to support bower which requires the extra files to be available in the git repo for the given tag.

So, with the fix in this PR, it makes it so that the commits are preserved. For example, in the currently released version of videojs on master, the [comparison link](https://github.com/videojs/video.js/blob/master/CHANGELOG.md#5123-2016-10-06) is from 5.10.7 to 5.12.3 because it didn't realize that there were other tags newer than 5.10.7. But it now picks up the correct previous tag.

Though, now, I've released 5.11.8 and when I'm trying to do a 5.12.4 release, I end up getting all the changelog from 5.12.3 because 5.11.8 comes out as the "latest" tag. Which it is, but I'm releasing 5.12, not 5.11. But because it's 5.11.8 comes out the latest tag, the 5.12.4 changelog continues to include the changelog from 5.12.3.

**Comment by stevemao** [comment]
@gkatsev Thanks. Could you open another issue for this :) Would be great if you could send a PR with a simple test for this :) So that the problem is easier to look at.

**Comment by gkatsev** [comment]
I'd be happy to.

Btw, here's a snippet of how the tags look like against npm with this change:

```
v3.10.9
v2.15.11
v3.10.8
v2.15.10
v3.10.7
v3.10.6
v3.10.5
v2.15.9
```

Also, I think in the meantime, I can manually add a context locally while we find the best solution to the problem.

File: test.js
Comment (JaKXz) [inline test.js:128]: @bcoe sorry for the late reply on this, my only nit here is: should we pass `err` to the `done` method so that it's not swallowed?

**Comment by stevemao** [comment]
Given this behaviour might break some users, I'm happy to add an option for this. Thoughts?

**Comment by bcoe** [comment]
@JaKXz @Tapppi @stevemao I'm not sure of the best approach here, sounds like we'll potentially introduce broken workflows regardless of which path we take -- maybe better that we stick with the existing logic, but document the fact that one should only generate their CHANGELOG from the `master` branch.

**Comment by gkatsev** [comment]
I think an option, disabled by default, would be preferable to not supporting this usecase at all.

**Comment by Tapppi** [comment]
I'm thinking an option too. Having considered this further I would prefer not to have this on by default in `standard-version` since it will screw up having multiple release branches, if I understood @gkatsev correctly. If and when that problem gets fixed somehow, it would make sense to have this on in `standard-version` too IMHO.

**Comment by Tapppi** [comment]
See https://github.com/conventional-changelog/standard-version/issues/130#issuecomment-264547521 for IMHO the nail in the coffin about whether this should be enabled in `standard-version`.  This can move forward as an option, disabled by default, IMO.

**Comment by stevemao** [comment]
@bcoe I really like this idea and would like to have it as an option (disabled by default). At least its useful for this module.

**Pull request closed without merging**
